src/lib.rs

```diff
@@ -40,7 +41,7 @@ pub use decode::Decoder;
 pub use encode::Encoder;
 
 /// The version number of the data structures.
-const FORMAT_VERSION: usize = 0;
+const SER_VERSION: usize = 0;
```

Comment:
I think this needs to be a fixed size integer, otherwise we might read in a file from another architecture and do the wrong thing. Which suggests we probably need to stick the architecture in here somewhere too...

Reply:
I think you are right on both counts. How about a u32 for the version, and the platform string?

Reply:
The platform string is variable size, which might be a bit annoying?

Reply:
I don't think the serialiser cares, but let's try it.

Failing that we can use a manually specified enum.

Reply:
Also (ho ho ho!) does the ser_version change alongside the VM version? In other words, we might keep the same serialisation format, but some other detail changes and then boom! So we probably need some sort of YK version number in there too?

Reply:
Maybe. Let me think about it.

Reply:
This is tricky, because ykrustc doesn't have any explicit knowledge about what it is compiling.

I think I agree that we need the version field though.

I wonder if we could add a flag or environment variable or something which ykrustc picks up.

Reply:
An alternative is simply not to worry about any of this. After all, if this stuff is compiled into the executable, then presumably everything in the executable is upto date, and there's nothing to worry about.

Reply:
Let's roll with that for now.

Reply:
OK, so we should remove the format version check and so on entirely then, I think.

Reply:
I've just understood you comment about it being built in to the executable.

The problem is, I *think*, that there are two compile-times: the compiler, and the VM. We could get into a scenario where the compiler was built with one version of ykpack, and the VM was built using another. Right?

So, is it the semantic version of ykpack we actually care for here? Should that be what is encoded?

Reply:
At least at first, we're probably going to compile to a big binary blob, so I don't think we need to worry about this (in other words, that blob will carry with it the run-time compiler which, by definition, must be in sync with everything else in the blob). I think we can just drop the version checks entirely, at least as long as we compile to a single blob.

Reply:
Hrm. I'm not sure I follow. I thought the VM would depend upon `ykrt` (and thus `ykpack`) via regular Cargo dependencies.

So the compiler will have (somewhere):
```
[dependencies]
ykpack = "x.y"
```

And then the VM would (perhaps transitively) have a similar dependency line in a Cargo manifest.

I've not considered the "blob" approach, nor know how hard it is to achieve.

Reply:
Exactly. cargo does the hard work for us, and we use the same tricks grmtools does to include small binary blobs into our big binary blob. If I'm wrong, we can put this stuff back later, but let's just drop the version checks for now, and make our life simpler.

Reply:
OK, I'll have to trust you on this one.

It seems to me that the VM author has to be careful to use the same version of `ykpack` as `ykrustc` used or boom.